### PR TITLE
core: fix race condition for TransportSet scheduleBackoff

### PR DESCRIPTION
**Pull request by dapengzhang0**
Trying to fix issue  #2188
- Try to keep avoiding the lock issue #2152 and also to avoid race condition #2188.
- Add `checkState` for `endBackoff()`. Could help hit and identify any potential issue related to #2188.
- Make sure `startBackoff()` and `endBackoff()` invoked in the right order.
- Not to schedule endBackoff if transportSet has been shutdown.

### CODE REVIEW

**Comment by ejona86** [comment]
OS X failed. I wonder if it just reproduced the race.

**Comment by dapengzhang0** [comment]
I think it is a failure caused by the assertion in `endBackoff()` @ejona86

File: core/src/main/java/io/grpc/internal/TransportSet.java [hunk 1]
```diff
@@ -280,16 +280,26 @@ public ClientTransport get() {
...
-    reconnectTask = scheduledExecutor.schedule(
-        new LogExceptionRunnable(new EndOfCurrentBackoff()), delayMillis, TimeUnit.MILLISECONDS);
     return new Runnable() {
       @Override
       public void run() {
         // This must be run outside of lock. The TransportSet lock is a channel level lock.
         // startBackoff() will acquire the delayed transport lock, which is a transport level
         // lock. Our lock ordering mandates transport lock > channel lock.  Otherwise a deadlock
         // could happen (https://github.com/grpc/grpc-java/issues/2152).
+        if (delayedTransport.isInBackoffPeriod()) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/TransportSet.java]: What flow allows getting here while already in backoff period?

Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/TransportSet.java]: this code shouldn't be there, was used to compare travis mac results, actually was removed in my draft

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java [hunk 2]
```diff
@@ -322,7 +322,11 @@ boolean isInBackoffPeriod() {
    */
   void startBackoff(final Status status) {
     synchronized (lock) {
-      Preconditions.checkState(backoffStatus == null);
+      if (shutdown) {
+        return;
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: not sure if we should add this

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java [hunk 3]
```diff
@@ -325,7 +325,8 @@ void startBackoff(final Status status) {
       if (shutdown) {
         return;
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: Not sure if this is correct

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java [hunk 4]
```diff
@@ -335,14 +339,17 @@ void startBackoff(final Status status) {
             it.remove();
           }
         }
-        streamCreationExecutor.execute(new Runnable() {
+
+        class FailTheFailFastPendingStreams implements Runnable {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: Why is this change needed?

File: core/src/main/java/io/grpc/internal/TransportSet.java [hunk 5]
```diff
@@ -280,16 +271,39 @@ public ClientTransport get() {
...
         // startBackoff() will acquire the delayed transport lock, which is a transport level
         // lock. Our lock ordering mandates transport lock > channel lock.  Otherwise a deadlock
         // could happen (https://github.com/grpc/grpc-java/issues/2152).
         delayedTransport.startBackoff(status);
+
+        synchronized (lock) {
+          if (shutdown) {
+            return;
+          }
+          scheduleEndOfCurrentBackoff();
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/TransportSet.java]: Are you trying to fix the issue? This should have been mentioned in the PR description.

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java (hunk 4 above)
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: it was used in the test, I'm not sure if there is a more convenient way in the test to run `startBackoff` and `TransportSet#shutdown` in the mean time.

File: core/src/main/java/io/grpc/internal/TransportSet.java (hunk 5 above)
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/TransportSet.java]: yes, added description.

File: core/src/main/java/io/grpc/internal/TransportSet.java [hunk 6]
```diff
@@ -233,15 +233,6 @@ private Runnable scheduleBackoff(
       final DelayedClientTransport delayedTransport, final Status status) {
     Preconditions.checkState(reconnectTask == null, "previous reconnectTask is not done");
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/TransportSet.java]: I think you need to move this check into scheduleEndOfCurrentBackoff(), which makes `scheduleBackoff()` do nothing other than returning a Runnable. This twisted monster is equivalent to its original form (before the changes made in #2154) called outside of synchronized in `transportShutdown()`.

Moving `scheduleBackoff()` outside of the synchronized block in `transportShutdown()` might be the better alternative to #2154. The lock in this case is to protect `reconnectPolicy` and `reconnectTask`, which are only changed in transport callbacks and when TransportSet is shutdown, thus nobody should change them between exiting the synchronized block and calling `scheduleBackoff()` in `transportShutdown()`.

File: core/src/main/java/io/grpc/internal/TransportSet.java [hunk 7]
```diff
@@ -227,21 +227,23 @@ private Runnable startNewTransport(DelayedClientTransport delayedTransport) {
...
-          new Object[]{getLogId(), delayMillis});
+    synchronized (lock) {
+      if (!shutdown) {
+        scheduleEndOfCurrentBackoff(delayedTransport);
+      }
     }
+  }
+
+  @GuardedBy("lock")
+  private void scheduleEndOfCurrentBackoff(final DelayedClientTransport delayedTransport) {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/TransportSet.java]: I would prefer this method to be merged back to `scheduleBackoff()`. Private methods exists for two reasons: to be reused, or to break up complicated task into smaller steps. Neither is the case here. Having a separate private method here only impedes code reading.

The `EndOfCurrentBackoff` definition doesn't need to be in the lock, so the synchronized block wouldn't be very long.

Comment (dapengzhang0) [inline core/src/main/java/io/grpc/internal/TransportSet.java]: I think it is the second case - although it's arguable.  I consider `scheduleEndofCurrentBackoff` or `scheduleBackoff` as a complicated task, if the former were merged into the latter, that's fine, then one usually adds a comment before it for better readability like `// schedule end of current backoff`, which is easily becoming obsolete.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/TransportSet.java]: scheduleBackoff() only has 4 lines of code. It's by no means complicated.

> ... then one usually adds a comment before it for better readability like `// schedule end of current backoff`

That's fine, although I think the code is self-descriptive enough so it can live without additional comment.

>... which is easily becoming obsolete.

Why?

**Comment by zhangkun83** [comment]
LGTM

**Comment by zhangkun83** [comment]
When you are going to merge it? I am also making changes in TransportSet and it may become harder for you to merge if done later.

**Comment by dapengzhang0** [comment]
rebased and squashed

**Comment by zhangkun83** [comment]
LGTM
